content/courses/advanced-ada/chapters/types.rst

```diff
@@ -1178,13 +1178,350 @@ declaration of the :ada:`Light` constant.
+In this sense, types have a constrained set of literals that can be used for
+objects of that type.
+
+User-defined literals allow us to extend this set of literals. We could, for
+example, extend the type declaration of :ada:`Activation_State` and allow the
+use of integer literals for objects of that type. In this case, we need to use
+the :ada:`Integer_Literal` aspect and specify a function that implements the
+conversion from literals to the type we're declaring. For this conversion from
+integer literals to the :ada:`Activation_State` type, we could specify that 0
+corresponds to :ada:`Off`, 1 corresponds to :ada:`On` and other values
+correspond to :ada:`Unknown`. We'll see the corresponding implementation later.
+
+.. note:: This feature was first introduced in Ada 2020 and might not be
+          available in older compilers.
+
+These are the three kinds of literals and their corresponding aspect:
+
++----------+----------+-------------------------+
+| Literal  | Example  | Aspect                  |
++==========+==========+=========================+
+| Integer  |        1 | :ada:`Integer_Literal`  |
++----------+----------+-------------------------+
+| Real     |      1.0 | :ada:`Real_Literal`     |
++----------+----------+-------------------------+
+| String   |     "On" | :ada:`String_Literal`   |
++----------+----------+-------------------------+
+
+For our previous :ada:`Activation_States` type, we could declare a function
+:ada:`Integer_To_Activation_State` that converts integer literals to one of the
+enumeration literals that we've specified for the :ada:`Activation_States`
+type:
+
+.. code:: ada manual_chop no_button project=Courses.Advanced_Ada.Types.User_Defined_Literals
+
+    !activation_states.ads
+    package Activation_States is
+
+       type Activation_State is (Unknown, Off, On)
+         with Integer_Literal => Integer_To_Activation_State;
+
+       function Integer_To_Activation_State (S : String)
+                                             return Activation_State;
+
+    end Activation_States;
+
+Based on this specification, we can now use an integer literal to initialize an
+object :ada:`S` of :ada:`Activation_State` type:
+
+.. code-block:: ada
+
+    S : Activation_State := 1;
+
+Note that we have a string parameter in the declaration of the
+:ada:`Integer_To_Activation_State` function, even though the function itself is
+only used to convert integer literals (but not string literals) to the
+:ada:`Activation_State` type. It's our job to process that string parameter in
+the implementation of the :ada:`Integer_To_Activation_State` function and
+convert it to an integer value |mdash| using :ada:`Integer'Value`, for example:
+
+.. code:: ada manual_chop compile_button project=Courses.Advanced_Ada.Types.User_Defined_Literals
+
+    !activation_states.adb
+    package body Activation_States is
+
+       function Integer_To_Activation_State (S : String)
```

Comment:
I think we should be consistent with the capitalization for the type name String. It is "string" in several places above, but also "String" too. (I'd advocate for "String".)